From dcb7c6ca80c40a2c8f1c07e32246091de4493aa2 Mon Sep 17 00:00:00 2001
From: Tamizh Chelvam Raja <quic_tamizhr@quicinc.com>
Date: Tue, 18 Jun 2024 00:15:25 +0530
Subject: [PATCH 1/2] QSDK: mac80211: Store netdev of AP_VLAN in ieee80211_sta
 structure

Store netdev pointer of AP_VLAN interface in ieee80211_sta structure
this will be used by driver in fast_rx handling.

Signed-off-by: Tamizh Chelvam Raja <quic_tamizhr@quicinc.com>
---
 include/net/mac80211.h | 1 +
 net/mac80211/cfg.c     | 6 ++++--
 2 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/include/net/mac80211.h b/include/net/mac80211.h
index b0456c8..7ac3921 100644
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -2487,6 +2487,7 @@ struct ieee80211_sta {
 
 	u16 eml_cap;
 	u16 mld_cap_op;
+	struct net_device *dev;
 
 	/* must be last */
 	u8 drv_priv[] __aligned(sizeof(void *));
diff --git a/net/mac80211/cfg.c b/net/mac80211/cfg.c
index 068b843..e48074a 100644
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -2423,12 +2423,14 @@ static int ieee80211_change_station(struct wiphy *wiphy,
 			rcu_assign_pointer(vlansdata->u.vlan.sta, sta);
 			__ieee80211_check_fast_rx_iface(vlansdata);
 
-			if (ieee80211_hw_check(&local->hw, SUPPORTS_NSS_OFFLOAD))
+			if (ieee80211_hw_check(&local->hw, SUPPORTS_NSS_OFFLOAD)) {
 				drv_sta_set_4addr(local, vlansdata, &sta->sta,
 						  true);
-			else
+			} else {
+				sta->sta.dev = vlansdata->dev;
 				drv_sta_set_4addr(local, sta->sdata, &sta->sta,
 						  true);
+			}
 			if (sta->sta.valid_links) {
 				int link_id;
 				for_each_set_bit(link_id,
-- 
2.34.1

